Source/Core/VideoCommon/FramebufferShaderGen.cpp

```diff
@@ -496,4 +521,126 @@ std::string GenerateFormatConversionShader(EFBReinterpretType convtype, u32 samp
+  ss << "{\n";
+  ss << "  int layer = int(v_tex0.z);\n";
+  ss << "  int4 coords = int4(int2(frag_coord.xy), layer, 0);\n";
+
+  // Convert to a 32-bit value encompassing all channels, filling the most significant bits with
+  // zeroes.
+  ss << "  uint raw_value;\n";
+  switch (from_format)
+  {
+  case TextureFormat::I8:
+  case TextureFormat::C8:
+  {
+    ss << "  float4 temp_value = ";
+    EmitTextureLoad(ss, 0, "coords");
+    ss << ";\n";
+    ss << "  raw_value = uint(temp_value.r * 255.0);\n";
+  }
+  break;
+
+  case TextureFormat::IA8:
+  {
+    ss << "  float4 temp_value = ";
+    EmitTextureLoad(ss, 0, "coords");
+    ss << ";\n";
+    ss << "  raw_value = uint(temp_value.r * 255.0) | (uint(temp_value.a * 255.0) << 8);\n";
+  }
+  break;
+
+  case TextureFormat::IA4:
+  {
+    ss << "  float4 temp_value = ";
+    EmitTextureLoad(ss, 0, "coords");
+    ss << ";\n";
+    ss << "  raw_value = uint(temp_value.r * 15.0) | (uint(temp_value.a * 15.0) << 4);\n";
+  }
+  break;
+
+  case TextureFormat::RGB565:
+  {
+    ss << "  float4 temp_value = ";
+    EmitTextureLoad(ss, 0, "coords");
+    ss << ";\n";
+    ss << "  raw_value = uint(temp_value.b * 31.0) | (uint(temp_value.g * 63.0) << 5) |\n";
+    ss << "              (uint(temp_value.r * 31.0) << 11);\n";
+  }
+  break;
+
+  case TextureFormat::RGB5A3:
+  {
+    ss << "  float4 temp_value = ";
+    EmitTextureLoad(ss, 0, "coords");
+    ss << ";\n";
+
+    // 0.8784 = 224 / 255 which is the maximum alpha value that can be represented in 3 bits
+    ss << "  if (temp_value.a > 0.878f) {\n";
+    ss << "    raw_value = (uint(temp_value.b * 31.0)) | (uint(temp_value.g * 31.0) << 5) |\n";
+    ss << "                (uint(temp_value.r * 31.0) << 10) | 0x8000u;\n";
+    ss << "  } else {\n";
+    ss << "     raw_value = (uint(temp_value.b * 15.0)) | (uint(temp_value.g * 15.0) << 4) |\n";
+    ss << "                 (uint(temp_value.r * 15.0) << 8) | (uint(temp_value.a * 7.0) << 12);\n";
+    ss << "  }\n";
+  }
+  break;
+  }
+
+  // Now convert it to its new representation.
+  switch (to_format)
+  {
+  case TextureFormat::I8:
+  case TextureFormat::C8:
+  {
+    ss << "  ocol0.rgba = (float(raw_value & 0xFFu) / 255.0).rrrr;\n";
+  }
+  break;
+
+  case TextureFormat::IA8:
+  {
+    ss << "  ocol0.rgb = (float(raw_value & 0xFFu) / 255.0).rrr;\n";
+    ss << "  ocol0.a = float((raw_value >> 8) & 0xFFu) / 255.0;\n";
+  }
+  break;
+
+  case TextureFormat::IA4:
+  {
+    ss << "  ocol0.rgb = (float(raw_value & 0xFu) / 15.0).rrr;\n";
+    ss << "  ocol0.a = float((raw_value >> 4) & 0xFu) / 15.0;\n";
+  }
+  break;
+
+  case TextureFormat::RGB565:
+  {
+    ss << "  ocol0 = float4(float((raw_value >> 10) & 0x1Fu) / 31.0\n";
+    ss << "                 float((raw_value >> 5) & 0x1Fu) / 31.0,\n";
+    ss << "                 float(raw_value & 0x1Fu) / 31.0,, 1.0);\n";
+  }
+  break;
+
+  case TextureFormat::RGB5A3:
+  {
+    ss << "  if ((raw_value & 0x8000u) != 0u) {\n";
```

Comment:
I'm not very well versed with the GC/Wii image formats, so I might be reading this wrong.  The contents inside this check seems like it is for a non alpha RGB5A3, is that correct?  It seems like we would be checking the top-bit and also checking for 1 instead of 0?